### ISSUE TITLE
TLS setup: do not chown if certificates have been not generated

### ISSUE THREAD

**Issue by EugenMayer** [osixia/docker-openldap#59]
Could you think about not chow-ning / fix permissions of certificates, if you did not generate them?
This could help using services like letsencrypt and others, when the certificates are shared to the host filesystem

**Comment by osixia** [comment]
Hello,
i will make some tests without chown-ing / fix permissions of certificates and let you know.

As i remember openldap was pretty strict with the certificates files permissions.

**Comment by EugenMayer** [comment]
thanks, let me know

**Comment by wrouesnel** [comment]
The solution I used for this on a derived image was to have the TLS certificates be copied on startup rather then used in place.

**Comment by osixia** [comment]
using something like `docker run [your options] osixia/openldap:1.1.3 --copy-service`  ?

**Comment by wrouesnel** [comment]
The old patch I used to use on this image just looked like [this](https://github.com/wrouesnel/docker-openldap/blob/4f813487bc68d61b874523ce1033eebd14e7d9ff/image/service/slapd/container-start.sh#L47)

A command line arg would be ideal, although is there any particular reason this shouldn't be the default? i.e. LDAP isn't going to be hot reloading the certificates is it?

**Comment by osixia** [comment]
The  --copy-service argument is an option from https://github.com/osixia/docker-light-baseimage 
it copy on start /container/service files to /container/run/service

In startup.sh CONTAINER_SERVICE_DIR is set to /container/run/service so all files used are the  /container/service files copies.

This should solve this issue.

**Comment by jgehrcke** [comment]
I'd also be in favor having the cerificate files copied by default, so that there is no write to that part of the host file system in any case.

Btw, I am using --copy-service but the container still changes file permissions of the cert files. Maybe I am invoking it the wrong way -- how can I check from the log that --copy-service actually took effect?

**Comment by osixia** [comment]
run `docker run osixia/openldap:1.1.3 --copy-service`

this should output : 
```
*** CONTAINER_LOG_LEVEL = 3 (info)
*** Copy /container/service to /container/run/service
*** Search service in CONTAINER_SERVICE_DIR = /container/run/service :
...
No certificate file and certificate key provided, generate:
/container/run/service/slapd/assets/certs/ldap.crt and /container/run/service/slapd/assets/certs/ldap.key
```

An other example with existing certificates : 
`docker run -v /data/ssl/ldap:/container/service/slapd/assets/certs -e LDAP_TLS_CA_CRT_FILENAME=ca.crt -e LDAP_TLS_CRT_FILENAME=cert.crt -e LDAP_TLS_KEY_FILENAME=cert.key osixia/openldap:1.1.3 --copy-service`

```
*** CONTAINER_LOG_LEVEL = 3 (info)
*** Copy /container/service to /container/run/service
*** Search service in CONTAINER_SERVICE_DIR = /container/run/service :
...
```
the cert mounted in /container/service/slapd/assets/certs/ must have been copied to /container/**run/**service/slapd/assets/certs/

The certificates in /container/**run/**service/slapd/assets/certs/ are then chmod + chown

**Comment by osixia** [comment]
After some tests openldap requires the certificate files to be owned by the **openldap** user.
If you mount these files from the host or any other volume and can't allow chmod and chown modification  (readonly volume for example) run the container with the command `--copy-service`.

Example: 
```
docker run -v /data/openldap/certs:/container/service/slapd/assets/certs [your other options] osixia/openldap --copy-service
```
What will append : 
On start the container will copy all files in **/container/service** to **/container/run/service**
and use only files from /container/run/service and ignore /container/service directory files. It's very important to keep mounting your files to /container/service/some_dir and not /container/run/service/some_dir otherwise --copy-service has no advantages.

I think using --copy-service by default would be realy confusing and make harder to understand what happens in the container.

**Comment by EugenMayer** [comment]
great option, thank you a lot. Does this "copying" happen on every container start ( not only creation )?

**Comment by osixia** [comment]
This is done only on container first start if the directory /container/run/service don't already exists.

https://github.com/osixia/docker-light-baseimage/blob/stable/image/tool/run#L372

https://github.com/osixia/docker-light-baseimage/blob/stable/image/tool/run#L421-L431

**Comment by EugenMayer** [comment]
well if one has to change his certificate, this could become a bit odd. Sure, one can remake the container, i guess that should not flush any data down the toilet i guess, right. Thats what is the idea behind the design right?

**Comment by osixia** [comment]
Yes, if the certificate needs to be updated, stop and remove the container and start a new one with the  same command and options.

**Comment by osixia** [comment]
This assumes that you have take care to mount the database (/var/lib/ldap) and config dir (/etc/ldap/slapd.d) to a volume to not lost any data.

**Comment by cochiseruhulessin** [comment]
I got this same issue with slapd trying to modify on readonly volumes in Kubernetes, but it was solved by using the `--copy-service` parameter.
